### ISSUE TITLE
Wrong type generated in the emitted type

### ISSUE THREAD

**Issue by PeterMK85** [astahmer/openapi-zod-client#208]
**Describe the bug**
Used version: `openapi-zod-client@1.10.5`

**Minimal reproduction**

<details>
    <summary>Contract</summary>
```yaml
openapi: 3.0.3
info:
    title: Example
    version: 1.0.0
tags:
  - name: pet
    description: Everything about your Pets
    externalDocs:
      description: Find out more
      url: http://swagger.io
paths:
  /pet:
    put:
      tags:
        - pet
      summary: Update an existing pet
      description: Update an existing pet by Id
      operationId: updatePet
      requestBody:
        description: Update an existent pet in the store
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/Pet"
          application/xml:
            schema:
              $ref: "#/components/schemas/Pet"
          application/x-www-form-urlencoded:
            schema:
              $ref: "#/components/schemas/Pet"
        required: true
      responses:
        "200":
          description: Successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Pet"
            application/xml:
              schema:
                $ref: "#/components/schemas/Pet"
        "400":
          description: Invalid ID supplied
        "404":
          description: Pet not found
        "405":
          description: Validation exception
      security:
        - petstore_auth:
            - write:pets
            - read:pets
components:
  schemas:
    Category:
      type: object
      properties:
        id:
          type: integer
          format: int64
          example: 1
        name:
          type: string
          example: Dogs
      xml:
        name: category
    Tag:
      type: object
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
      xml:
        name: tag
    Pet:
      required:
        - name
        - photoUrls
      type: object
      properties:
        id:
          type: integer
          format: int64
          example: 10
        name:
          nullable: true
          example: doggie
          oneOf:
            - type: string
            - type: number
        category:
          $ref: "#/components/schemas/Category"
        photoUrls:
          type: array
          xml:
            wrapped: true
          items:
            type: string
            xml:
              name: photoUrl
        tags:
          type: array
          xml:
            wrapped: true
          items:
            $ref: "#/components/schemas/Tag"
        status:
          type: string
          description: pet status in the store
          enum:
            - available
            - pending
            - sold
      xml:
        name: pet
    ApiResponse:
      type: object
      properties:
        code:
          type: integer
          format: int32
        type:
          type: string
        message:
          type: string
      xml:
        name: "##default"
  requestBodies:
    Pet:
      description: Pet object that needs to be added to the store
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/Pet"
        application/xml:
          schema:
            $ref: "#/components/schemas/Pet"
  securitySchemes:
    petstore_auth:
      type: oauth2
      flows:
        implicit:
          authorizationUrl: https://petstore3.swagger.io/oauth/authorize
          scopes:
            write:pets: modify pets in your account
            read:pets: read your pets
    api_key:
      type: apiKey
      name: api_key
      in: header
</details>

Used with the option of `--export-types`

### Issue itself:
```ts
type Pet = {
    id?: number | undefined
    name: string | number
    category?: Category | undefined
    photoUrls: Array<string>
    tags?: Array<Tag> | undefined
    status?: ('available' | 'pending' | 'sold') | undefined
}

const Pet: z.ZodType<Pet> = z.object({
    id: z.number().int().optional(),
    name: z.union([z.string(), z.number()]).nullable(),
    category: Category.optional(),
    photoUrls: z.array(z.string()),
    tags: z.array(Tag).optional(),
    status: z.enum(['available', 'pending', 'sold']).optional(),
})
```

Highlight of the issue:

Contract:
```yaml
        name:
          nullable: true
          example: doggie
          oneOf:
            - type: string
            - type: number
```

generated Type:
```ts
name: string | number
```

But it should be:
```
name: string | number | null
```

In the generated type

**Expected behavior**
The `Pet` type should match with the actual returned type

**Comment by astahmer** [comment]
@eli0shin maybe you want to take a loot at it

**Comment by eli0shin** [comment]
I don't think that this is new, I have run into small mismatches between the typescript and zod output in the past.
I switched my template to apply the type with as to unblock the issue but that's not correct, it just unblocked me temporarily.
I will dig into the open api to typescript code today to see if there is an issue with how allOf/anyOf are handled.

**Comment by astahmer** [comment]
btw the minimal repro is in the wrong format, can you format it and post it again ? or save it in the playground https://openapi-zod-client.vercel.app/ (cmd+s or actions > save)

**Comment by PeterMK85** [comment]
@astahmer 

I followed, but I think there is an issue with the [preview](https://openapi-zod-client.vercel.app/?doc=PYBwpgdghiCWBcACAzAOgAyuQKFhAZsPNoogC6xkA2YSAygO5QDmzYATogAphkDOZYOzCIAtIgDy4CAEEuASRQYSiACZg%2BAY3awQFYBCQAfUStIAVABaw%2BiG4iiI%2BUALYga3XojqDh3jgBuHIgARlB8YKqIBuSWIlKQcopo6E7gmrD4sJpQ%2BhCopACawACuiDkQiFkQUaVkiC5CIlAhdWbeTKzBuYgA2pZkZCB88AD0o3ydbOyosMAAugAUA0Mj45Ms07PAAJQF8pVkcbGw7FGUHLlzlcD4sSLg9QJNADSIDGAA5EFODJSacSignuag0sGYlSy7AEDhAIHYwCgAIAhO1imUKogIMAGIg4lQQIgSrZYG4ET8jiIku84pTOJRPrYQgBPBpQADWeGY5UsUAgbFswMpoOqlGudn4YCodyE5GAIM0wHUqHaVh6TGZbz%2BR3IpLAWpEmNJ8OAFOO1LwiDYEEuVDefKiYAAHiBgBEnMAXCJbiCbbj8GBciVhCTKhIZHQ0Nh2gB9eR3ZmlT5%2BKjAYCc-lVWXCxibYIAJgwo3DdEQhdSQWh4p9PH4vn190qmio2XZfTiwiWK2GY1GkUoQlQGy6MzmowA-MGqABebtrUaPZ7CIdTDjbUYBfMTVczZmuKh7RAyKhkDjQChBKiaxCJjF8xCpqBRAKwRzCgAGAFFVJREAA%2BRAABlESiWsl3iCMywwd8GkgMpQDyZEY2jUhSDoT0RGJMB8BKKgHzwdkRnacReisERa28etEGEV0%2BAHdhmS7QYe3GZhKEsEoQlQRUXG3PN2FEGBYD4kdREXesdmIvoyI9YNNCpBQRTwMUYkIThhQonwmiY1ZezYo5OO4z0ROmQS4BMjgxN4cDRhCVMQlGFxwlPdgJnYTRHKgPBRhDUp3I0UZQEgITUD3FwDxUFyXD4CR8DoQJsloPFmN7YctjHKK%2BFGFRFQgMgkTIYhUMQMAnNgKgkCE09XAAATStc5hUFt5IgCIitQ6AvSQGQQCRY5y3aKckG7XsGDG1AYD6sBUCEZhRmayAIiywD5AAYU-AA5OhP1EctUAGcKVErOiDCQABGDBUDOs7sHy5giNIcROqSx52nULQdD0a4kE-StmSOLkHFaEp6lvTgwPaZ0XOgKgABFgE0B7itBD7dDyJAADE8FqEGGiadpSCG5KhlSndthUJ7XCS8C3o0bQ0e%2Bo9NHkvhBXlMCqKEdRoUhp1oagOGEaR4r3vpr7TsQLGamiXHGj8Fo6hlzgaeRol2Aq4mQFJ-jycerEqaQLD2Fp1HxcMSRwHYK4DFsBXcaN7BeqOJGF14drSBAEH3dQu7heR8RXtVvgShcJyGKQABVEBVFyZpKmdGwKEzQPkdFz70cQKOY9PBx46dRPAceUJWXkVQCeiS3rYgUvDej2Pa3L4QAEcSg0MgACElWZb2Rbp9PGaz2Pc5K-OBEgeoi8tYUVdV8oDFPPKe%2BRmB3GyKvRgAKz4U7y6DgFSqgJfZ8QAASYR8CQAAiABiUYeNdG08qyrQ4icrLa0v3fUJX5r16dcKj7IxfgfQBqsz7YSvrfe%2BBhx7P33m-UYH8v6kB-mvPIownSiDGgwUQakXCiCnJARU6hVCgNQsApyZDirgIvogG%2Bd9PQP1gRMeB4REG8E-sfZuJRTiRCQGQdgrdG4aAfktQBl98zoHQJfMhacGYSzoCUZmGg%2BA4TwkFK2eRkFzzyuPKhsJV45HQVvHex896v0PtosB59IEMLcDAp%2BLCLHvw4VYgxv90H-wqm40gFDLFmNnjQ2x0DH78CcQfFxZBOGz0vgAFikTI7RcizZIAOAEAWsAojyFhk4EocIWyRC-nE9AsTEkBOSRnCi2J6iEBKDUIp8SACsZSzEVMZgANQydncUzp5Jm3LhETQwZKDdy-gHay9YYxQBBpYfR4gGA6FPPARcbjxDCCfMs6y2AQmwPan4v2q1Y7MCEKM1WZBmTgCQMAEIG8wCaDIOXE0lsKAaEAZksh5zLl2F0dMbReDchIDwGQAAbLE7RzpXDuCSjdY%2Bz0PkXOpgIrk4KnSQpoEgeG91y5eMAXC8oRyTmqhYEvT5SVrm3PuY8hEzzYCvK-u87RpLAU-I4H8oQTlCrfJBWC2FBtGUIqQAIHQ-JsUAK-niu67RaxL24bw0hYz9ZegVSASwwBBAR3VnwcuTLog3LuQ81WTyOAvL9sVBlASdVArAL8gJ-zOVAtBSitF0L0Dir5RagVTgkUioCRCtw6K1DAFYLSr%2BRjrUnLIUEuhUDGEOLCfs0YhzTzHIYtEw1qr1WavhV8qA7ArbMm0TitxCyV58PIIIsA2iLjRX0TqoVyKAmkCLY2jqBtEAqrVcADVVAv6%2B2zUlXN%2BbC1ipbSWuEZaBFCICdW01gSbHRrsUwxxCbzAsDTUA-KZBiT9sFd65gSS%2B7yPNkXAQQZQwghnsfOCLg5kOHSeVFoNBVntsgD%2BH1jbxDbyoGXVWzbZ54pTqQHqsAABKIiba0G1Z68l%2BqqUaJNYA4hkGPVfKtTasxdrmVkGQPmXtAr%2BVfPre%2B1WXpWYsGQ2Yute7RXeN5V1aN191D4GmSeaJ3C26dx-HS4q0ry5tIlhRGD9zYg9BtJENmoRmiqBIXKC99Zy65QXoVL%2BqCjHXE3tvQw2i-H6KjfQnZy7WGRPXcVVTf8R1mJ024vTMb7GhLgc49hUSgA).

Issues with the Preview:
![image](https://github.com/astahmer/openapi-zod-client/assets/3931377/5241242d-7964-48d6-8364-4d9c9bd4b791)
Adds undefined.

![image](https://github.com/astahmer/openapi-zod-client/assets/3931377/f12ecd88-afd5-499c-900d-e9ecb7e267dd)
Even when picks up properly the types not reflected on the preview.

**Comment by astahmer** [comment]
good catch ! but that's another issue.

can you paste your yaml from your OP
![Screenshot 2023-08-18 at 16 24 20](https://github.com/astahmer/openapi-zod-client/assets/47224540/1825509c-1d66-4da2-80c0-ade0fe18da59)


but the usual format:
![Screenshot 2023-08-18 at 16 24 45](https://github.com/astahmer/openapi-zod-client/assets/47224540/8847ad7d-9b3e-4dbd-aacc-3e2e204ceb60)

cause that playground link doesn't seem to be the same schema, I can't find "oneOf" in there

**Comment by PeterMK85** [comment]
@astahmer 

Sure 👍 

Let's retry the `[preview](https://openapi-zod-client.vercel.app/?doc=PYBwpgdghiCWBcACAzAOgAyuQKFhAZsPNoogC6xkA2YSAygO5QDmzYATogAphkDOZYOzCIAtIgDy4CAEEuASRQYSiACZg%2BAY3awQFYBCQAfUStIAVABaw%2BiG4iiI%2BUALYga3XojqDh3jgBuHIgARlB8YKqIBuSWIlKQcopo6E7gmrD4sJpQ%2BhCopACawACuiDkQiFkQUaVkiC5CIlAhdWbeTKzBuYgA2pZkZCB88AD0o3ydbOyosMAAugAUA0Mj45Ms07PAAJQF8pVkcbGw7FGUHLlzlcD4sSLg9QJNADSIDGAA5EFODJSacSignuag0sGYlSy7AEDhAIHYwCgAIAhO1imUKogIMAGIg4lQQIgSrZYG4ET8jiIku84pTOJRPrYQgBPBpQADWeGY5UsUAgbFswMpoOqlGudn4YCodyE5GAIM0wHUqHaVh6TGZbz%2BR3IpLAWpEmNJ8OAFOO1LwiDYEEuVDefKiYAAHiBgBEnMAXCJbiCbbj8GBciVhCTKhIZHQ0Nh2gB9eR3ZmlT5%2BKjAYCc-lVWXCxibYIAJgwo3DdEQhdSQWh4p9PH4vn190qmio2XZfTiwiWK2GY1GkUoQlQGy6MzmowA-MGqABebtrUaPZ7CIdTDjbUYBfMTVczZmuKh7RAyKhkDjQChBKiaxCJjF8xCpqBRAKwRzCgAGAFFVJREAA%2BRAABlESiWsl3iCMywwd8GkgMpQDyZEY2jUhSDoT0RGJMB8BKKgHzwdkRnacReisERa28etEGEV0%2BAHdhmS7QYe3GZhKEsEoQlQRUXG3PN2FEGBYD4kdREXesdmIvoyI9YNNCpBQRTwMUYkIThhQonwmiY1ZezYo5OO4z0ROmQS4BMjgxN4cDRhCVMQlGFxwlPdgJnYTRHKgPBRhDUp3I0UZQEgITUD3FwDxUFyXD4CR8DoQJsloPFmN7YctjHKK%2BFGFRFQgMgkTIYhUMQMAnNgKgkCE09XAAATStc5hUFt5IgCIitQ6AvSQGQQCRY5y3aKckG7XsGDG1AYD6sBUCEZhRmayAIiywD5AAYU-AA5OhP1EctUAGcKVErOiDCQABGDBUDOs7sHy5giNIcROqSx52nULQdD0a4kE-StmSOLkHFaEp6lvTgwPaZ0XOgKgABFgE0B7itBD7dDyJAADE8FqEGGiadpSCG5KhlSndthUJ7XCS8C3o0bQ0e%2Bo9NHkvhBXlMCqKEdRoUhp1oagOGEaR4r3vpr7TsQLGamiXHGj8Fo6hlzgaeRol2Aq4mQFJ-jycerEqaQLD2Fp1HxcMSRwHYK4DFsBXcaN7BeqOJGF14drSBAEH3dQu7heR8RXtVvgShcJyGKQABVEBVFyZpKmdGwKEzQPkdFz70cQKOY9PBx46dRPAceUJWXkVQCeiS3rYgUvDej2Pa3L4QAEcSg0MgACElWZb2Rbp9PGaz2Pc5K-OBEgeoi8tYUVdV8oDFPPKe%2BRmB3GyKvRgAKz4U7y6DgFSqgJfZ8QAASYR8CQAAiABiUYeNdG08qyrQ4icrLa0v3fUJX5r16dcKj7IxfgfQBqsz7YSvrfe%2BBhx7P33m-UYH8v6kB-mvPIownSiDGgwUQakXCiCnJARU6hVCgNQsApyZDirgIvogG%2Bd9PQP1gRMeB4REG8E-sfZuJRTiRCQGQdgrdG4aAfktQBl98zoHQJfMhacGYSzoCUZmGg%2BA4TwkFK2eRkFzzyuPKhsJV45HQVvHex896v0PtosB59IEMLcDAp%2BLCLHvw4VYgxv90H-wqm40gFDLFmNnjQ2x0DH78CcQfFxZBOGz0vgAFikTI7RcizZIAOAEAWsAojyFhk4EocIWyRC-nE9AsTEkBOSRnCi2J6iEBKDUIp8SACsZSzEVMZgANQydncUzp5Jm3LhETQwZKDdy-gHay9YYxQBBpYfR4gGA6FPPARcbjxDCCfMs6y2AQmwPan4v2q1Y7MCEKM1WZBmTgCQMAEIG8wCaDIOXE0lsKAaEAZksh5zLl2F0dMbReDchIDwGQAAbLE7RzpXDuCSjdY%2Bz0PkXOpgIrk4KnSQpoEgeG91y5eMAXC8oRyTmqhYEvT5SVrm3PuY8hEzzYCvK-u87RpLAU-I4H8oQTlCrfJBWC2FBtGUIqQAIHQ-JsUAK-niu67RaxL24bw0hYz9ZegVSASwwBBAR3VnwcuTLog3LuQ81WTyOAvL9sVBlASdVArAL8gJ-zOVAtBSitF0L0Dir5QEiAuEqAtHReQQRYAnVuF9aoYArBaXaJgbFOZ5ABVOCRSKgJesdWepcCEVlx8jHWpOWQoJdCoGMIcWE-ZoxDmnmOQxaJhrVXqs1fCr5UB2BW2ZNonFbiFkrz4X6oRASLjRX0TqoVyLE2IFbcO0geKVVquABqqgX9fZ1qSg2ptLaxXDvbXCTtAju1mN7aawJNi812KYY44t5gWCVqAflMgxIF2CvjcwJJfd5HmyLgIIMoYQQz2PnBFw0aoDpPKj6gNw7xk1CHYm8Q28qBl1VqO2eE7eDtB6rAAASiIm2tBtWxvJfqqlGiTWAOIZhi1sarU2rMXa5lZBkD5jnQK-lXzB0Jtnl6VmLBiNmIHfe0V3jeVdTzdfdQ%2BBpknmidwtuncfx0uKtK8ubSJYURw-c2IPQbSRDZqEZoqgSFyk-fWcuuUF6FS-qgox1xN7b0MNovx%2Bjc30J2Se1hkSL3FVM3-VdZibNuLs-m%2BxoS4HOPYVEoAA&prettier=EQbwOgdgBDVsAHATgSwgFwOooCboBbwBcUAjAEwAMANJLHMOgIYBG2ehwJALLdLPCZIkAewDuABSEBTCAGdiDJgBsxTAJ4LgfevACuc6QBVWWkgDMVhnQOCGAtikXoke6TZjw5aAObLpAIp6IujSipbK1nS2Lkwoyr4AwiL29kyK8NJyAKzwkAC%2BwEA)`


As inlined:
```yaml
openapi: 3.0.3
info:
  title: Swagger Petstore - OpenAPI 3.0
  description: |-
    This is a sample Pet Store Server based on the OpenAPI 3.0 specification.  You can find out more about
    Swagger at [https://swagger.io](https://swagger.io). In the third iteration of the pet store, we've switched to the design first approach!
    You can now help us improve the API whether it's by making changes to the definition itself or to the code.
    That way, with time, we can improve the API in general, and expose some of the new features in OAS3.

    _If you're looking for the Swagger 2.0/OAS 2.0 version of Petstore, then click [here](https://editor.swagger.io/?url=https://petstore.swagger.io/v2/swagger.yaml). Alternatively, you can load via the `Edit > Load Petstore OAS 2.0` menu option!_

    Some useful links:
    - [The Pet Store repository](https://github.com/swagger-api/swagger-petstore)
    - [The source API definition for the Pet Store](https://github.com/swagger-api/swagger-petstore/blob/master/src/main/resources/openapi.yaml)
  termsOfService: http://swagger.io/terms/
  contact:
    email: apiteam@swagger.io
  license:
    name: Apache 2.0
    url: http://www.apache.org/licenses/LICENSE-2.0.html
  version: 1.0.11
tags:
  - name: pet
    description: Everything about your Pets
    externalDocs:
      description: Find out more
      url: http://swagger.io
  - name: store
    description: Access to Petstore orders
    externalDocs:
      description: Find out more about our store
      url: http://swagger.io
  - name: user
    description: Operations about user
paths:
  /pet:
    put:
      tags:
        - pet
      summary: Update an existing pet
      description: Update an existing pet by Id
      operationId: updatePet
      requestBody:
        description: Update an existent pet in the store
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/Pet"
          application/xml:
            schema:
              $ref: "#/components/schemas/Pet"
          application/x-www-form-urlencoded:
            schema:
              $ref: "#/components/schemas/Pet"
        required: true
      responses:
        "200":
          description: Successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Pet"
            application/xml:
              schema:
                $ref: "#/components/schemas/Pet"
        "400":
          description: Invalid ID supplied
        "404":
          description: Pet not found
        "405":
          description: Validation exception
      security:
        - petstore_auth:
            - write:pets
            - read:pets
components:
  schemas:
    Category:
      type: object
      properties:
        id:
          type: integer
          format: int64
          example: 1
        name:
          type: string
          example: Dogs
      xml:
        name: category
    Tag:
      type: object
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
      xml:
        name: tag
    Pet:
      required:
        - name
        - photoUrls
      type: object
      properties:
        id:
          type: integer
          format: int64
          example: 10
        name:
          nullable: true
          example: doggie
          oneOf:
            - type: string
            - type: number
        category:
          $ref: "#/components/schemas/Category"
        photoUrls:
          type: array
          xml:
            wrapped: true
          items:
            type: string
            xml:
              name: photoUrl
        tags:
          type: array
          xml:
            wrapped: true
          items:
            $ref: "#/components/schemas/Tag"
        status:
          type: string
          description: pet status in the store
          enum:
            - available
            - pending
            - sold
      xml:
        name: pet
    ApiResponse:
      type: object
      properties:
        code:
          type: integer
          format: int32
        type:
          type: string
        message:
          type: string
      xml:
        name: "##default"
  requestBodies:
    Pet:
      description: Pet object that needs to be added to the store
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/Pet"
        application/xml:
          schema:
            $ref: "#/components/schemas/Pet"
```
